### DISCUSSION TITLE
Will there be Linux support available?

### DISCUSSION THREAD

**Discussion by I3ladeDragon** [tryashtar/nbt-studio#9] [Q&A]
I use Ubuntu 20.04 and remember needing Wine to install the old NBTExplorer.

I LOVED that program and wish to use this one! 

That said, Wine causes conflicts with a program we use for a web interface on my Bungeecord network.

Even if I cannot use this, I love that the ol' NBTExplorer is getting a reboot! <3

**Comment by tryashtar** [comment]
Hey there. Thanks for your post. I would love to support Linux but unfortunately, I have no experience with that kind of thing and don't really have the means right now to learn/develop/test it. So without help or anything, I'm sorry to say this will not be a priority for a good while

**Reply by I3ladeDragon** [reply]
Sounds good! Hopefully someone randomly decides they feel the need to help port it sometime! hahaha! Keep up the good work in the meantime!

**Comment by MestreLion** [comment]
Care to check if the following was done? https://stackoverflow.com/questions/47285243/

The 3 top voted answers seem very promising...

**Reply by tryashtar** [reply]
This is winforms, not a console project. It won't be able to run without wine

**Reply by MestreLion** [reply]
NTB Explorer is not a console project either, and I can run it with `mono` (and its winforms package) without wine.

There is also `dotnet` for Linux, provided by Microsoft itself, that allows building and running .NET applications.

Is it possible to make NBT Studio compatible with either `mono` or `dotnet`? Or, if `wine` is really required, what would be the needed "addons" (instalable via Winetricks) to make it work?

**Reply by tryashtar** [reply]
Yeah, NBT Explorer uses mono. Sadly I have no experience with mono

**Reply by MestreLion** [reply]
NBT Explorer does not "use" mono, it's kinda the other way around: NBT Explorer is a .NET application (just like NBT Studio), and mono can _run_ .NET applications. But NBT Explorer itself is completely unaware of mono.

Both Explorer and Studio were developed in .NET (and both uses Winforms), so in theory `mono` should be able to run both. And so should `dotnet`.

There is something in Studio's setup, configuration, or components that makes mono and dotnet unable to run it out-of-the-box. The link I posted has some promising suggestions on what needs to be configured to make it run fine on Linux. Care to check if there is any useful/relevant information there?

**Reply by tryashtar** [reply]
Oh you're right, what I remembered about NBTEx turned out to just be Mac-specific stuff to make it native-ish.
If you're getting an error or something when attempting to use mono, let me know and I'll check it out. But no promises.

**Reply by MestreLion** [reply]
Thanks!!! I've made several attempts using wine, mono, dotnet, when rushing trying to run it with. I'll do it in a more methodological way so we can investigate the issue.

Meanwhile, my googling showed that ".NET" is actually 2 very distinct APIs, with very distinct cross-platform support:
- .NET Framework, older and full-fledged, which `mono` handles quite nicely. Deprecated by Microsoft (circa... 2018? 16? 20?)
- .NET Core, newer, slimmer, replaced the above around version 5. MS itself provides cross-platform via `dotnet` but for fewer components on Non-Windows. Poorly (or not at all) supported by `mono`

So, what exact framework / components / platform / etc NBT Studio is or uses? What are its dependencies?

I can feed you with any tests required on the Linux side, thanks for being open to this!

**Reply by tryashtar** [reply]
Yeah, sorry I was not very clear about that. NBT Studio uses .NET 5 (so "core," although that name was dropped as framework became obsolete). This is mostly an arbitrary choice on my part; I updated the project from framework, but not for any particular reason besides newer=better.
I was able to run it through wine on my Linux Fedora VM, only after installing a couple random .NET things from Microsoft.

**Reply by MestreLion** [reply]
Yeah, I believe `mono` can't run .NETCore/.NET5 software, only 4.x Framework, and its successor, `dotnet` from Microsoft, has no Winforms on Linux (and it seems it won't ever have, not even on .NET6 when `mono` will be merged to `dotnet`. Sad.). Microsoft is promoting a new GUI API to replace Winforms (MUIA or something), but A) Developers would have to port their code; B) There's no guarantee that this new API will be available to linux at launch.

So, for now it's either reverting back to 4.x Framework for the sake of `mono` compatibility (would that require much/any code change?), at least for 1-2 years to see how cross-platform .NET6 will really be (or a better solution arises), or use `wine` with a given set of .NET addons (care to document which ones you had to install to make it run in your Fedora?)

**Reply by MestreLion** [reply]
.NET ~~Core~~ (or simply .NET 5+) is the future and the road ahead, no question about it. _But_... that's very new technology with very little cross-compatibility for GUIs/Desktop apps such as NBT Studio. Even for _Windows_ Winforms was only available in .NET Core since version 3, in the end of **2019**!! That's very little time for support and compatibility spread and settle.

**Reply by tryashtar** [reply]
I'm pretty sure it was this link: https://dotnet.microsoft.com/download/dotnet/thank-you/runtime-desktop-5.0.11-windows-x64-installer

I don't know why C# doesn't seem to have any good UI technologies. Sadly, I only know winforms, but I've been trying to learn WPF. Even still, that is also windows-only, at least officially. I don't know anything about mono/wine.

I don't really want to revert the .NET 5 port. I'm very sorry about all of this.

**Comment by MestreLion** [comment]
I was able to successfully run it under `wine`!

Here are the needed steps:
- Get a _really_ modern `wine`. The one in Ubuntu's 18.04 repositories is not enough. I installed latest wine 6.0.2 from the wine repository using the [official instructions](https://wiki.winehq.org/Ubuntu)
- Download the [.NET 5 Runtime for **Desktop Apps**, _**Windows**_ version, x64](https://dotnet.microsoft.com/download/dotnet/5.0/runtime) and install it _using_ wine (`wine64 windowsdesktop-runtime-*-win-x64.exe`)
- Run NBT Studio with wine: `wine64 ./NbtStudio.exe`

Have fun!

**Comment by MestreLion** [comment]
I also wrote a simple installer / runner with all the needed commands, for a semi-automated installer. It sets up the wine repository and dependencies using commands adapted from the official sources:

```bash
#!/bin/bash

# https://github.com/tryashtar/nbt-studio

########################################################
# Install Wine 6 (maybe 5 works too, who knows?)
# Check if already integrated with seten

# https://wiki.winehq.org/Ubuntu
sudo dpkg --add-architecture i386

### if Ubuntu 18.04, need extra packages from OpenSuse. They're already available in 20.04
wget -qO- 'https://download.opensuse.org/repositories/Emulators:/Wine:/Debian/xUbuntu_18.04/Release.key' | sudo apt-key add -
sudo apt-add-repository --no-update 'deb https://download.opensuse.org/repositories/Emulators:/Wine:/Debian/xUbuntu_18.04/ ./'
####

wget -qO- 'https://dl.winehq.org/wine-builds/winehq.key' | sudo apt-key add -
sudo apt-add-repository --no-update 'https://dl.winehq.org/wine-builds/ubuntu/'
sudo apt update

sudo apt install -y --install-recommends winehq-stable
##########################################################

# This is optional, but I prefer to keep my wine "bottles" isolated from each other
export WINEPREFIX=$HOME/.local/share/wineprefixes/nbtstudio

# .NET 5 (formerlly .NET Core (up to 3.x), successor of .NET Framework (up to 4.x))
# https://dotnet.microsoft.com/download
# Windows, .NET 5.0 (or later), .NET Runtime
# https://dotnet.microsoft.com/download/dotnet/5.0/runtime
# Windows, Run Desktop Apps (not Console or Server), Download x64

# Automation:
# https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script

# Create and configure the bottle before installing .NET
wine64 wineboot 

# Install the .NET Runtime
wine64 windowsdesktop-runtime-*-win-x64.exe

###################
# Download and extract latest NBT Studio
# https://github.com/tryashtar/nbt-studio/releases/latest

wine64 ./NbtStudio.exe
```

@tryashtar , would you be interested in a PR to include such script in NBT Studio? I could give it a better polish, add some sanity checks and verifications, auto-select the needed requirements (i.e., check if Ubuntu 18.04 or later), add a CLI for `--help`, etc...

**Reply by tryashtar** [reply]
That is awesome!
